tephra-core/src/main/java/co/cask/tephra/visibility/ReadBarrier.java

```diff
@@ -0,0 +1,79 @@
+/*
+ * Copyright © 2015 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.tephra.visibility;
+
+import co.cask.tephra.Transaction;
+import co.cask.tephra.TransactionAware;
+import com.google.common.primitives.Bytes;
+import com.google.common.primitives.Longs;
+
+import java.util.Collection;
+import java.util.Collections;
+
+/**
+ * Implementation of read barrier used by {@link VisibilityBarrier}.
+ */
+class ReadBarrier implements TransactionAware {
+  private final byte[] prefix;
+  private final byte[] barrierId;
+  private Transaction tx;
+
+  public ReadBarrier(byte[] prefix, byte[] barrierId) {
+    this.prefix = prefix;
+    this.barrierId = barrierId;
+  }
+
+  @Override
+  public void startTx(Transaction tx) {
+    this.tx = tx;
+  }
+
+  @Override
+  public void updateTx(Transaction tx) {
+    // Barriers need original write pointer
+  }
+
+  @Override
+  public Collection<byte[]> getTxChanges() {
+    if (tx == null) {
+      throw new IllegalStateException("Transaction has not started yet");
+    }
+    return Collections.singleton(Bytes.concat(prefix, Longs.toByteArray(tx.getWritePointer()), barrierId));
```

Comment:
Why ``tx.getWritePointer()`` instead of ``tx.getTransationId()``? I understand its the same in some cases but it might make it easier to understand?